src/expression.d

```diff
@@ -10852,7 +10852,29 @@ public:
         TypeVector tv = cast(TypeVector)tb;
         Type te = tv.elementType();
         dim = cast(int)(tv.size(loc) / te.size(loc));
-        return this;
+
+        bool checkElem(Expression elem)
+        {
+            if (elem.op == TOKint64 ||
+                elem.op == TOKfloat64 ||
+                elem.op == TOKcomplex80)
+                return false;
+
+             error("constant expression expected, not %s", elem.toChars());
+             return true;
+        }
+
+        e1 = e1.optimize(WANTvalue);
+        bool result;
+        if (e1.op == TOKarrayliteral)
+        {
+            foreach (i; 0 .. dim)
+                result |= checkElem((cast(ArrayLiteralExp)e1).getElement(i));
```

Comment:
> Then a comment would be in order.

added

> Actually this should just use  e.isConst() == 1

Good catch, done

>  and duplicate the error message call, no need for a nested function.

Using the nested function to factor out duplicated code makes the logic a bit simpler. It's a style that's ugly in C++, but D makes it natural. I intend to use it more often.